### PR TITLE
WIP hugolib: Add template analysis feature

### PR DESCRIPTION
**Pull request by moorereason**
This commit adds a templateAnalysis option to turn on tracking of
execution stats for referenced templates.

### CODE REVIEW

**Comment by moorereason** [comment]
This is a proof of concept.  Assuming it can be improved upon.  I'm not sure how to move the timings map up higher into the build state so that we can track the parent templates that are executed in `hugolib`.

File: tpl/tplimpl/template.go [hunk 1]
```diff
@@ -62,6 +63,9 @@ type GoHTMLTemplate struct {
 	amberFuncMap template.FuncMap
 
 	*deps.Deps
+
+	sync.RWMutex
```
Comment (bep) [inline tpl/tplimpl/template.go]: As you never use the read lock, a regular Mutex is a better choice.

Comment (moorereason) [inline tpl/tplimpl/template.go]: True.  I was using RLock in an earlier iteration.

**Comment by moorereason** [comment]
Refactored to push the timings up into `deps.Deps`.  This may be an abuse of the Deps concept, but I'm now able to track all template executions.

**Comment by bep** [comment]
It is a very good idea, and, with one important addition, would have saved me lots of debugging time the other day (I ended up adding something similar temporarily).

But you are right, it does not belong in `Deps`.

I suggest that this is important enough for its own package ( `metrics`?) and a simpler to use abstraction (`Timer.Time(func(), keys ...string)`?).

Also add a `Count` to the metrics (invocation count per key(s)).

The above probably means an adjustment of how the template methods are used (must consolidate the lookup and execute), but that seems worth doing.

I also suggest that the `GoHTMLTemplate` owns its own metrics, and also provides the metric report on demand (maybe implementing an optional interface with a `PrintMetrics(io.Writer)` method.

**Comment by bep** [comment]
Note: I'm about to simplify the Template interface to be able to do both text and html template, so it might be worth it to wait a little.

**Comment by moorereason** [comment]
Replaced by #3812

**Pull request closed without merging**